host/xtest/xtest_1000.c

```diff
@@ -1017,3 +1030,96 @@ static void xtest_tee_test_1012(ADBG_Case_t *c)
 		TEEC_CloseSession(&session1);
 	}
 }
+
+struct test_1013_thread_arg {
+	TEEC_SharedMemory *shm;
+	uint32_t error_orig;
+	TEEC_Result res;
+	uint32_t max_concurrency;
+};
+
+static void *test_1013_thread(void *arg)
+{
+	struct test_1013_thread_arg *a = arg;
+	TEEC_Session session = { 0 };
+	TEEC_Operation op = TEEC_OPERATION_INITIALIZER;
+	static const uint8_t sha256_in[] = { 'a', 'b', 'c' };
+	uint8_t out[32] = { 0 };
+
+	a->res = xtest_teec_open_session(&session, &concurrent_ta_uuid, NULL,
+					 &a->error_orig);
+	if (a->res != TEEC_SUCCESS)
+		return NULL;
+
+	op.params[0].memref.parent = a->shm;
+	op.params[0].memref.size = a->shm->size;
+	op.params[0].memref.offset = 0;
+	op.params[1].value.a = 10000;
+	op.params[1].value.b = 0;
+	op.params[2].tmpref.buffer = (void *)sha256_in;
+	op.params[2].tmpref.size = sizeof(sha256_in);
+	op.params[3].tmpref.buffer = out;
+	op.params[3].tmpref.size = sizeof(out);
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_PARTIAL_INOUT,
+					 TEEC_VALUE_INOUT,
+					 TEEC_MEMREF_TEMP_INPUT,
+					 TEEC_MEMREF_TEMP_OUTPUT);
+
+	a->res = TEEC_InvokeCommand(&session, TA_CONCURRENT_CMD_SHA256, &op,
+				    &a->error_orig);
+	a->max_concurrency = op.params[1].value.b;
+	TEEC_CloseSession(&session);
+	return NULL;
+}
+
+static void xtest_tee_test_1013(ADBG_Case_t *c)
+{
+	size_t num_threads = 5;
+	size_t n;
+	pthread_t thr[num_threads];
+	struct test_1013_thread_arg arg[num_threads];
+	size_t max_concurrency = 0;
+	TEEC_SharedMemory shm;
+
+	Do_ADBG_BeginSubCase(c, "Test concurrency with %zu parallel threads",
+				num_threads);
+
+	memset(&shm, 0, sizeof(shm));
+	shm.size = sizeof(struct ta_concurrent_shm);
+	shm.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		TEEC_AllocateSharedMemory(&xtest_teec_ctx, &shm)))
+		goto out;
+	memset(shm.buffer, 0, shm.size);
+
+	memset(arg, 0, sizeof(arg));
+	for (n = 0; n < num_threads; n++) {
+		arg[n].shm = &shm;
+		if (!ADBG_EXPECT(c, 0, pthread_create(thr + n, NULL,
+						test_1013_thread, arg + n)))
+			num_threads = n; /* break loop and start cleanup */
+	}
+
+	for (n = 0; n < num_threads; n++) {
+		(void)ADBG_EXPECT(c, 0, pthread_join(thr[n], NULL));
+		(void)ADBG_EXPECT_TEEC_SUCCESS(c, arg[n].res);
+		if (arg[n].max_concurrency > max_concurrency)
+			max_concurrency = arg[n].max_concurrency;
+	}
+
+	TEEC_ReleaseSharedMemory(&shm);
+
+	Do_ADBG_Log("Max concurrency %zu", max_concurrency);
+	/*
+	 * Concurrency can be limited by several factors, for instance in a
+	 * single CPU system it's dependent on the Preemtion Model used by
+	 * the kernel (Preemptible Kernel (Low-Latency Desktop) gives the
+	 * best result there).
+	 */
+	(void)ADBG_EXPECT_COMPARE_UNSIGNED(c, max_concurrency, >, 0);
```

Comment:
Should be `>, 1` otherwise you're not checking that at least some level of concurrency was possible.
Maybe this test should be skipped when optee_os was built with `CFG_NUM_THREADS = 1`?